### PR TITLE
Improve testing when multiple built-in providers are enabled

### PR DESCRIPTION
**Pull request by ctz**
Prior to this PR, `cargo test --all-features` runs the various unit/integration tests with one arbitrary provider and leaves the other untested. After this PR, `cargo test --all-features` runs tests that depend on a provider twice (once for aws-lc-rs, once for ring).

I'm not super happy about using macros for this. An alternative would be https://crates.io/crates/rstest which uses proc-macros to enable parameterised tests and fixtures.

### CODE REVIEW

File: rustls/tests/api.rs [hunk 1]
```diff
@@ -5736,56 +5683,112 @@ fn test_client_removes_tls12_session_if_server_sends_undecryptable_first_message
...
-    assert!(client_config.fips());
-    client_config.require_ems = false;
-    assert!(!client_config.fips());
+} // test_for_each_provider!
+
+use std::cell::RefCell;
+
+thread_local!(static COUNTS: RefCell<LogCounts> = RefCell::new(LogCounts::new()));
+
+struct CountingLogger;
```
Comment (djc) [inline rustls/tests/api.rs:5730]: Feels like this could/should be a separate commit.

**Review by djc — APPROVED** [review]

File: rustls/tests/api.rs (hunk 1 above)
Comment (ctz) [reply rustls/tests/api.rs:5730]: have moved this to the bottom of the file in a prior commit

File: rustls/tests/api.rs [hunk 2]
```diff
@@ -1316,52 +1329,46 @@ impl ResolvesClientCert for ClientCheckCertResolve {
     }
 }
 
+fn default_signature_schemes(version: ProtocolVersion) -> Vec<SignatureScheme> {
```
[the remark this answers is not in the document]
Comment (ctz) [reply rustls/tests/api.rs:1332]: moved down

File: rustls/tests/api.rs [hunk 3]
```diff
@@ -5736,56 +5683,112 @@ fn test_client_removes_tls12_session_if_server_sends_undecryptable_first_message
     ));
 }
 
-#[cfg(all(feature = "ring", not(feature = "fips")))]
 #[test]
 fn test_client_fips_service_indicator() {
-    assert!(!make_client_config(KeyType::Rsa).fips());
+    if provider_is_fips() {
```
[the remark this answers is not in the document]
Comment (ctz) [reply rustls/tests/api.rs:5688]: replaced with a straightforward `assert_eq!` against `provider_is_fips()`

**Comment by cpu** [comment]
Another unfortunate side effect of this change is that CLion is now unable to pick up the `#[test]` annotated functions within the macros as unit tests, removing the gutter-icon for running a single test with a click. It's possible to work around of course, but just wanted to note another bit of friction :'(

**Comment by ctz** [comment]
Ah, ew. I wonder if the rstest route would work with that integration? I guess it is based on something the language server tells the IDE?

**Comment by cpu** [comment]
> Ah, ew. I wonder if the rstest route would work with that integration?

I cloned the `rstest` repo and it doesn't seem to work with the `#[rstest]` decorated tests either. 

>  I guess it is based on something the language server tells the IDE?

I'm not sure but that seems like a reasonable assumption.
